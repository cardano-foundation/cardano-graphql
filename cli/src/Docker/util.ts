import { QuestionCollection } from 'inquirer'
import { DockerComposeStack, StatefulService } from './index'

export function selectServicesFromStack (stack: DockerComposeStack, multi = true): QuestionCollection {
  const label = multi === true ? 'services' : 'service'
  return {
    name: label,
    message: `Select ${label}`,
    type: multi === true ? 'checkbox' : 'list',
    choices: [...stack.statefulServices.values()].map(service => {
      return {
        name: service.name,
        value: service
      }
    }),
    validate (input) {
      if (input.length === 0) {
        throw new Error('At least one service must be selected')
      }
      return true
    }
  }
}

export function servicesFromNameArg (stack: DockerComposeStack, names: string): StatefulService[] {
  return names.split(',').map((name: string) => {
    if (!stack.statefulServices.has(name)) {
      throw new Error(`Unknown service ${name}`)
    }
    return stack.statefulServices.get(name)
  })
}
