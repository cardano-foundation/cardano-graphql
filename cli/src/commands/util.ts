import { DockerComposeStack } from '../docker'

export function serviceChoices (stack: DockerComposeStack) {
  return [...stack.statefulServices.values()].map(service => {
    return { name: service.name, value: service }
  })
}

export function serviceFromName (stack: DockerComposeStack, name: string) {
  if (!stack.statefulServices.has(name)) {
    throw new Error(`Unknown service ${name}`)
  }
  return stack.statefulServices.get(name)
}

export function servicesFromNames (stack: DockerComposeStack, names: string) {
  return names.split(',').map((name: string) => serviceFromName(stack, name))
}
