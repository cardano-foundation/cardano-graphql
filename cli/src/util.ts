
export function ensureValue (input: string) {
  return !input ? 'Cannot be blank' : true
}
