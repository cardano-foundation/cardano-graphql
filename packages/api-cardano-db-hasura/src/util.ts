import { ConnectionConfig, createInteractionContext } from '@cardano-ogmios/client'
import { Logger } from 'ts-log'

export const createInteractionContextWithLogger = (connection: ConnectionConfig, logger: Logger, module: string, onClose?: () => Promise<void>) =>
  createInteractionContext(
    (error) => {
      logger.error({ module, error }, error.message)
    },
    async (code, reason) => {
      if (code === 1000) {
        logger.info({ module, code }, reason)
      } else {
        logger.error({ module, code }, 'Connection closed')
        await onClose?.()
      }
    },
    {
      connection
    }
  )
