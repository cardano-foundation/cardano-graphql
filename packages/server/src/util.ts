import { Application } from 'express'
import http from 'http'

export function listenPromise (app: Application, port: number, hostname?: string): Promise<http.Server> {
  return new Promise(function (resolve, reject) {
    const server: http.Server = app.listen(port, hostname, () => resolve(server))
    server.on('error', reject)
  })
}
