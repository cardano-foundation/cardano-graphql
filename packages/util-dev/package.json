{
  "name": "@cardano-graphql/util-dev",
  "version": "1.0.0",
  "description": "Common development utilities",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "scripts": {
    "build": "tsc -b ./src",
    "cleanup": "shx rm -rf dist node_modules",
    "lint": "eslint --ignore-path ../../.eslintignore \"**/*.ts\"",
    "prepack": "yarn build",
    "test": "shx echo No tests in this package"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/input-output-hk/cardano-graphql.git"
  },
  "author": "Rhys Bartels-Waller",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/input-output-hk/cardano-graphql/issues"
  },
  "homepage": "https://github.com/input-output-hk/cardano-graphql/blob/master/packages/util-dev/README.md",
  "devDependencies": {
    "@cardano-graphql/util": "1.0.0",
    "shx": "^0.3.2"
  },
  "directories": {
    "lib": "src",
    "test": "test"
  },
  "files": [
    "dist",
    "LICENSE",
    "NOTICE",
    "README.md"
  ]
}
