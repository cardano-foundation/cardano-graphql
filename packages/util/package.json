{
  "name": "@cardano-graphql/util",
  "version": "1.0.0",
  "description": "Common utilities",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "scripts": {
    "build": "tsc -b ./src",
    "cleanup": "shx rm -rf dist node_modules",
    "lint": "eslint --ignore-path ../../.eslintignore \"**/*.ts\"",
    "prepack": "yarn build",
    "test": "shx echo No tests in this package"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/input-output-hk/cardano-graphql.git"
  },
  "author": "Rhys Bartels-Waller",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/input-output-hk/cardano-graphql/issues"
  },
  "homepage": "https://github.com/input-output-hk/cardano-graphql/blob/master/packages/util/README.md",
  "dependencies": {
    "graphql": "14.5.8",
    "graphql-bigint": "^1.0.0",
    "graphql-tag": "^2.10.3",
    "p-retry": "^4.2.0"
  },
  "devDependencies": {
    "@types/node": "^14.0.13",
    "shx": "^0.3.2",
    "typescript": "^3.9.5"
  },
  "directories": {
    "lib": "src",
    "test": "test"
  },
  "files": [
    "dist",
    "LICENSE",
    "NOTICE",
    "README.md"
  ]
}
