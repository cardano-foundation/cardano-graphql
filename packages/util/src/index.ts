import { onFailedAttemptFor } from './onFailedAttemptFor'
import { loadQueryNode } from './queryNodeLoading'
import * as dateTimeValidation from './dateTimeValidation'
import * as scalars from './scalars'

export default {
  dateTimeValidation,
  onFailedAttemptFor,
  loadQueryNode,
  scalars
}
