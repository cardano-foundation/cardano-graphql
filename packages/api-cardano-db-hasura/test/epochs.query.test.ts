import gql from 'graphql-tag'
import path from 'path'

import { DocumentNode } from 'graphql'
import util from '@cardano-graphql/util'
import { TestClient } from '@cardano-graphql/util-dev'
import { epoch1 } from './data_assertions'
import { buildClient } from './util'

function loadQueryNode (name: string): Promise<DocumentNode> {
  return util.loadQueryNode(path.resolve(__dirname, '..', 'src', 'example_queries', 'epochs'), name)
}

describe('epochs', () => {
  let client: TestClient
  beforeAll(async () => {
    client = await buildClient()
  }, 60000)

  it('Returns epoch details by number', async () => {
    const result = await client.query({
      query: await loadQueryNode('epochDetailsByNumber'),
      variables: { number: 1 }
    })
    expect(result.data.epochs[0]).toEqual(epoch1.basic)
    expect(result.data).toMatchSnapshot()
  })

  it('Can return aggregated data', async () => {
    const result = await client.query({
      query: await loadQueryNode('aggregateDataWithinEpoch'),
      variables: { number: 1 }
    })
    expect(result.data.epochs[0]).toEqual(epoch1.aggregated)
    expect(result.data.epochs[0].blocksCount).toEqual(epoch1.aggregated.blocks_aggregate.aggregate.count)
    expect(result.data).toMatchSnapshot()
  })

  it('Can return filtered aggregated data', async () => {
    const result = await client.query({
      query: await loadQueryNode('numberOfBlocksProducedByLeaderInEpoch'),
      variables: { number: 1, slotLeader: 'SlotLeader-6c9e14978b9d6629' }
    })
    expect(result.data).toMatchSnapshot()
  })

  it('Returns epoch details by number range', async () => {
    // Todo: Convert this into an actual ranged query now the performance issue is resolved.
    const result = await client.query({
      query: await loadQueryNode('epochDetailsInRange'),
      variables: { numbers: [1] }
    })
    expect(result.data.epochs[0]).toEqual(epoch1.basic)
    expect(result.data).toMatchSnapshot()
  })

  it('Can return aggregated Epoch data', async () => {
    const result = await client.query({
      query: await loadQueryNode('aggregateEpochData'),
      variables: { epochNumberLessThan: 185 }
    })
    expect(result.data).toMatchSnapshot()
  })

  it('Returns blocks scoped to epoch', async () => {
    const validQueryResult = await client.query({
      query: await loadQueryNode('blocksInEpoch'),
      variables: { number: 1, blockLimit: 1 }
    })
    const invalidQueryResult = await client.query({
      query: gql`query {
          epochs( where: { number: { _eq: 1 }}) {
              blocks(limit: 20, where: { epoch: { number: { _eq: 0 } }}) {
                  hash
              }
          }
      }`
    })
    expect(validQueryResult.data.epochs[0].blocks[0].epoch.number).toBe(1)
    expect(invalidQueryResult.data.epochs[0].blocks.length).toBe(0)
  })
})
