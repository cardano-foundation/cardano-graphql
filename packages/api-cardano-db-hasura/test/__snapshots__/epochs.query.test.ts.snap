// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`epochs Can return aggregated Epoch data 1`] = `
Object {
  "epochs_aggregate": Object {
    "aggregate": Object {
      "count": "185",
      "max": Object {
        "number": 184,
        "output": "424616166073756860",
        "transactionsCount": "40972",
      },
      "min": Object {
        "output": "2654692166740613",
        "transactionsCount": "33",
      },
      "sum": Object {
        "output": "8293029495565297000",
        "transactionsCount": "1938088",
      },
    },
  },
}
`;

exports[`epochs Can return aggregated data 1`] = `
Object {
  "epochs": Array [
    Object {
      "blocksCount": "21589",
      "blocks_aggregate": Object {
        "aggregate": Object {
          "avg": Object {
            "fees": 47848.56538051786,
            "size": 838.0591968131919,
          },
          "count": "21589",
          "max": Object {
            "fees": "1377042",
            "size": "4787",
          },
          "min": Object {
            "fees": "0",
            "size": "631",
          },
          "sum": Object {
            "fees": "1033002678",
            "size": "18092860",
          },
        },
      },
      "number": 1,
    },
  ],
}
`;

exports[`epochs Can return filtered aggregated data 1`] = `
Object {
  "epochs": Array [
    Object {
      "blocks_aggregate": Object {
        "aggregate": Object {
          "count": "3123",
        },
      },
      "number": 1,
    },
  ],
}
`;

exports[`epochs Returns epoch details by number 1`] = `
Object {
  "epochs": Array [
    Object {
      "blocksCount": "21589",
      "lastBlockTime": "2017-10-03T21:43:51.000Z",
      "number": 1,
      "output": "101402912214214220",
      "startedAt": "2017-09-28T21:45:51.000Z",
      "transactionsCount": "12870",
    },
  ],
}
`;

exports[`epochs Returns epoch details by number range 1`] = `
Object {
  "epochs": Array [
    Object {
      "blocksCount": "21589",
      "lastBlockTime": "2017-10-03T21:43:51.000Z",
      "number": 1,
      "output": "101402912214214220",
      "startedAt": "2017-09-28T21:45:51.000Z",
      "transactionsCount": "12870",
    },
  ],
}
`;
